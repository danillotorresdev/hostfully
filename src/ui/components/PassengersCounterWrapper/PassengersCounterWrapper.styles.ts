import { styled } from "@linaria/react";

export const Label = styled.label`
  margin-bottom: 0.2rem;
  font-size: 0.8rem;
`;

export const Container = styled.div`
  position: relative;
  flex-direction: column;
  color: #333333;
`;

export const PassengerCounterWrapper = styled.div`
  position: absolute;
  top: 20px;
  width: 100%;
  z-index: 100;
`;

export const PassengerButton = styled.button`
  padding: 0 10px;
  border: none;
  cursor: pointer;
  border-radius: 10px;
  display: flex;
  align-items: center;
  width: 100%;
  background-color: #fff;
  border: 1px solid #ccc;
  height: 57px;
  font-size: 14px;
  .svg {
    margin-right: 5px;
  }
  &:hover {
    border: 1px solid #2d2aa5;
  }
`;
