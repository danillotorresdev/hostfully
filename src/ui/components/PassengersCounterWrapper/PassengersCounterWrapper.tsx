import { useRef, useState } from "react";
import { Icon } from "@/ui/components/Icon/Icon";
import * as S from "./PassengersCounterWrapper.styles";
import {
  PassengerCount,
  PassengerCounter,
} from "@/ui/components/PassengersCounterWrapper/PassengersCounter/PassengersCounter";
import { useClickoutside } from "@/utils/hooks/useClickOutside/useClickoutside";

type PassengerCounterProps = {
  onChange: (count: PassengerCount) => void;
  defaultValue?: PassengerCount;
};

export const PassengerCounterInput = ({
  onChange,
  defaultValue,
}: PassengerCounterProps) => {
  const [showCounter, setShowCounter] = useState(false);
  const [passengerCount, setPassengerCount] = useState<PassengerCount>(
    defaultValue ?? { adults: 1, children: 0 },
  );

  const counterRef = useRef<HTMLDivElement>(null);

  useClickoutside({
    ref: counterRef,
    callback: () => setShowCounter(false),
  });

  return (
    <S.Container>
      <S.Label>Passengers</S.Label>
      <S.PassengerButton
        type="button"
        onClick={() => setShowCounter(!showCounter)}
      >
        <Icon name="member" />
        {passengerCount.adults + passengerCount.children} passenger{passengerCount.adults + passengerCount.children > 1 ? "s" : ""}
      </S.PassengerButton>
      {showCounter && (
        <S.PassengerCounterWrapper>
          <PassengerCounter
            defaultValue={passengerCount}
            onChange={(count) => {
              onChange(count);
              setPassengerCount(count);
              setShowCounter(false);
            }}
            ref={counterRef}
          />
        </S.PassengerCounterWrapper>
      )}
    </S.Container>
  );
};
