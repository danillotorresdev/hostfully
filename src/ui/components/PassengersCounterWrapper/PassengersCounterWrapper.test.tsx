import { PassengerCounterInput } from "./PassengersCounterWrapper";
import { describe, expect, it, vi } from "vitest";
import { render, fireEvent, screen } from "@/utils/settings/tests/utilities";

describe("PassengerCounterWrapper", () => {
  it("should show the correct number of passengers", () => {
    const onChangeMock = vi.fn();
    const { asFragment } = render(
      <PassengerCounterInput
        onChange={onChangeMock}
        defaultValue={{
          adults: 1,
          children: 0,
        }}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText("1 passenger")).toBeDefined();
  });

  it('calls onChange with correct passenger counts when "Select" button is clicked', () => {
    const onChangeMock = vi.fn();
    const { getByText } = render(
      <PassengerCounterInput
        defaultValue={{
          adults: 1,
          children: 0,
        }}
        onChange={onChangeMock}
      />,
    );

    const passenger = getByText("1 passenger");

    fireEvent.click(passenger);

    const adultsIncrement = getByText("Adults")
      .nextElementSibling as HTMLButtonElement;
    const getAdultMinusButton = adultsIncrement.querySelectorAll("button")[0];
    const getAdultPlusButton = adultsIncrement.querySelectorAll("button")[1];
    fireEvent.click(getAdultPlusButton);
    fireEvent.click(getAdultPlusButton);

    fireEvent.click(getAdultMinusButton);

    const childrenIncrement = getByText("Children")
      .nextElementSibling as HTMLButtonElement;

    fireEvent.click(childrenIncrement.querySelectorAll("button")[1]);
    fireEvent.click(childrenIncrement.querySelectorAll("button")[1]);

    fireEvent.click(childrenIncrement.querySelectorAll("button")[0]);

    const selectButton = getByText("Select");

    fireEvent.click(selectButton);

    expect(onChangeMock).toHaveBeenCalledWith({ adults: 2, children: 1 });
  });
});
