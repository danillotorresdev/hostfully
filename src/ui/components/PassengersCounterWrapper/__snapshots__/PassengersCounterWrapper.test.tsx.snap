// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PassengerCounterWrapper > should show the correct number of passengers 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-1 c1u669jx"
  >
    <label
      class="mocked-styled-0 lzhktur"
    >
      Passengers
    </label>
    <button
      class="mocked-styled-3 pfu8a7v"
      type="button"
    >
      <svg
        class="svg svg-member"
        data-testid="icon-svg"
        style="width: 24px; height: 24px;"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="/assets/sprite.svg#member"
        />
      </svg>
      1 passenger
    </button>
  </div>
</DocumentFragment>
`;
